window-copy.c

```diff
@@ -3097,27 +3152,97 @@ window_copy_search(struct window_mode_entry *wme, int direction, int regex,
 	wrapflag = options_get_number(wp->window->options, "wrap-search");
 	cis = window_copy_is_lowercase(str);
 
-	if (direction)
+	keys = options_get_number(wp->window->options, "mode-keys");
+
+	if (direction) {
+		/*
+		 * Behave according to mode-keys. If it is emacs, search forward
+		 * leaves the cursor after the match. If it is vi, the cursor
+		 * remains at the beginning of the match, regardless of
+		 * direction, which means that we need to start the next search
+		 * after the term the cursor is currently on when searching
+		 * forward.
+		 */
+		if (keys == MODEKEY_VI) {
+			if (data->searchmark != NULL)
+				window_copy_move_after_search_mark(data, &fx,
+				    &fy, wrapflag);
+			else {
+				/*
+				 * When there are no search marks, start the
+				 * search after the current cursor position.
+				 */
+				window_copy_move_right(s, &fx, &fy, wrapflag);
+			}
+		}
 		endline = gd->hsize + gd->sy - 1;
+	}
 	else {
-		if (again)
-			window_copy_move_left(s, &fx, &fy, wrapflag);
+		/* Scan to just before current search mark. */
+		if (data->searchmark != NULL)
+			window_copy_move_before_search_mark(data, &fx, &fy,
+			    wrapflag);
 		endline = 0;
 	}
 
 	found = window_copy_search_jump(wme, gd, ss.grid, fx, fy, endline, cis,
-	    wrapflag, direction, regex, &foundlen);
+	    wrapflag, direction, regex);
 	if (found) {
 		window_copy_search_marks(wme, &ss, regex, visible_only);
-		if (foundlen != 0) {
-			/* Adjust for wrapped lines eating one right. */
-			i = data->cx + foundlen;
-			while (i > gd->sx - 1) {
-				i -= gd->sx;
-				window_copy_cursor_right(wme, 1);
+		fx = data->cx;
+		fy = screen_hsize(data->backing) - data->oy + data->cy;
+
+		/*
+		 * When searching forward, if the cursor is not at the beginning
+		 * of the mark, search again.
+		 */
+		if (direction &&
+		    window_copy_search_mark_at(data, fx, fy, &at) == 0 &&
+		    at > 0 &&
+		    data->searchmark[at] == data->searchmark[at - 1]) {
+			window_copy_move_after_search_mark(data, &fx, &fy,
+			    wrapflag);
+			window_copy_search_jump(wme, gd, ss.grid, fx,
+			    fy, endline, cis, wrapflag, direction,
+			    regex);
+			fx = data->cx;
+			fy = screen_hsize(data->backing) - data->oy + data->cy;
+		}
+
+		if (direction) {
+			/*
+			 * When in Emacs mode, position the cursor just after
+			 * the mark.
+			 */
+			if (keys == MODEKEY_EMACS) {
+				window_copy_move_after_search_mark(data, &fx,
+				    &fy, wrapflag);
+				data->cx = fx;
+				data->cy = fy - screen_hsize(data->backing) +
+				    data-> oy;
+			}
+		}
+		else {
+			/*
+			 * When searching backward, position the cursor at the
+			 * beginning of the mark.
+			 */
+			if (window_copy_search_mark_at(data, fx, fy,
```

Comment:
I tried to extract this too but ran into a strange bug I couldn't track down or consistently reproduce on my first reverse search (the cursor would end up after the mark, rather than at the beginning)

Reply:
Interesting. I combined the "move_before_start" and "move_to_start" functions and it seems to work fine. Please feel free to test the attached diff and incorporate the change if it works for you. The changed function is `window_copy_move_before_search_mark`. I will test this further with your latest diff.
[a.diff.txt](https://github.com/tmux/tmux/files/6176775/a.diff.txt)

Reply:
This diff doesn't use fx/fy to change the cursor position. I also don't know the code well enough to know what happens if you move before the 0th position in the grid and then move right again with wrap off. Would it end up on the 2nd character of the mark?

And I'm personally not a fan of passing a boolean to a function to fundamentally change its behavior, I see that as two functions. I'm good leaving this as-is, if you are.

Reply:
Sounds good. I'm ok with this being as it is now.